[WIP] Swift 3.0

marcelofabri opened the pull request:
This adds support to Swift 3.0. It's still WIP:
- [x] Update core
- [x] Update core tests
- [x] Update RAC
- [x] Update RAC tests
- [ ] Update podspec
- [x] Point OHHTTPStubs to https://github.com/AliSoftware/OHHTTPStubs/pull/201 or wait until that is merged
- [x] ~~Update tests to run on iOS 10: they seem to be failing on the Simulator with the same issue mentioned here: https://forums.developer.apple.com/thread/51071 and http://stackoverflow.com/questions/38456471/secitemadd-always-returns-error-34018-in-xcode-8-beta-4-in-ios-10-simulator~~ I think this should be done on a later PR, since it'll probably need to create an app target so the unit tests could be run on a host app (with an Entitlements file)
- [ ] Lock dependencies to releases (some of them are referring to branches)

marcelofabri commented on Heimdallr/Core/URLRequestExtensions.swift:
```diff
@@ -109,45 +107,13 @@ public extension NSMutableURLRequest {
-                let substring = string.substringWithRange(range)
-
-                escaped += substring.stringByAddingPercentEncodingWithAllowedCharacters(allowedCharacterSet) ?? substring
+        var allowedCharacterSet = CharacterSet.urlQueryAllowed
```
I've updated this from Alamofire code. It seems that they are dropping iOS 8.x on the new release. Maybe we should do the same? (If not, this needs to be reverted and converted to Swift 3.0)

hffmnn replied:
From my point of view it is ok to drop iOS 8 support. If someone needs it, we are open for a Pull Request :)

marcelofabri commented:
Also, I chose to not mark the classes as `open` as the converter suggested, but that may be discussible

marcelofabri commented:
I think Circle hasn't update to the GM yet

marcelofabri commented on Cartfile.resolved:
```diff
@@ -1,6 +1,6 @@
+github "antitypical/Result" "3.0.0"
 github "jspahrsummers/xcconfigs" "0.9"
-github "ReactiveCocoa/ReactiveCocoa" "v4.2.2"
+github "ReactiveCocoa/ReactiveCocoa" "v4.0.0-alpha.3"
```
for some reason, this was changed when I updated Nimble

¯\_(ツ)_/¯

marcelofabri commented:
So, the Mac and iOS builds are passing, but the Reactive Cocoa isn't. From what I've seen, 4.x won't support Swift 3, so we'd need to wait for 5.0. I have no idea if `master` is currently stable or not as I have little familiarity with RC.

What do you guys think?

SusannProszak commented:
hello @marcelofabri 
thanks for your pr. 👍 
we prefer to wait until ReactiveCocoa releases swift 3.0 support. Until then I would leave the pr in [WIP].

judi0713 commented:
So, is there any plan to migrate Heimdallr to Swift 3.0?
I already use Xcode 8😔Recently I want to update my project to support swift 3.0.

marcelofabri commented:
Should we use https://github.com/ReactiveCocoa/ReactiveSwift instead of RAC 4? If so, there's already a 1.0.0-alpha.1 release we can use.

marcelofabri commented:
I've updated the dependencies now that Nimble and Quick have stable versions and also updated the code to use ReactiveSwift instead of ReactiveCocoa. 

However, I had to comment all the `Objective-C compatibility` section from `Heimdallr+ReactiveCocoa` because I didn't know how to migrate that.

BabyAzerty commented:
Excuse me, I don't have much details into the project but as a user of Heimdallr, I would like to understand why is there - historically - blocking dependencies still here?
There was already a problem with the JSON parser before (Argo).
And now RAC.

Or is it that we can DL this branch, exclude RAC, and get things working?

marcelofabri commented:
@BabyAzerty If you use CocoaPods, you can use my fork without the RAC subspec (it's optional), so you can migrate your project to Swift 3.0 while waiting this PR to be merged. Just change your Podfile:

```ruby
pod 'Heimdallr', git: 'https://github.com/marcelofabri/Heimdallr.swift.git', branch: 'swift-3.0'
```

While I can't speak for the maintainers, I think it'd be useful to make the RAC extension another repository to make it easier to manage, but I don't know how complex that would make things.

bckr commented:
> I've updated the dependencies now that Nimble and Quick have stable versions and also updated the code to use ReactiveSwift instead of ReactiveCocoa.

Thank you so much for this and sorry for our late response here. We've been a bit busy the last couple of weeks but I would like to rekindle the discussion here again.

> However, I had to comment all the Objective-C compatibility section from Heimdallr+ReactiveCocoa because I didn't know how to migrate that.

I think if we want to have full compatibility with Objc and Swift we'll require both `ReactiveCocoa` and the `ReactiveObjCBridge` now because the repository [will be split up](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3118/files) into multiple smaller ones for the upcoming release that brings Swift 3 compatibility.

> While I can't speak for the maintainers, I think it'd be useful to make the RAC extension another repository to make it easier to manage, but I don't know how complex that would make things.

I like that idea! I believe it would make the project way easier to maintain. Since we are also internally moving away from heavy RAC usage in our project, it might be the right direction for `Heimdallr` as well. The reactive extensions can still live in a separate repository. I'll discuss it internally and comment here if we have made a final decision. Otherwise the Swift 3 migration would be blocked until we get a stable RAC 5.0 release.

bckr commented:
We decided that we are fine with splitting up the repository and removing RAC as a dependency for this repository.

@marcelofabri Do you want to tackle that? I think you already provide a working Swift 3 fork without the RAC subspec? We'll create a second repository for the reactive extensions when we have a stable RAC release. Let me know if I can help you out in any way 😊

marcelofabri commented:
@bckr I'm currently traveling, so it's unlikely that I'm able to work on this for the next ~2 weeks. The fork currently have a RAC subspec, but it's just not working 100%.

I've allowed edits from maintainers, so feel free to start from my fork if you wanna handle this.

bckr commented:
@marcelofabri Will do! Thanks!

Ingibjorg commented:
Hi all,
what is the ETD on this update? Heimdallr is the only library that prevents me from being able to update my project from Swift 2.3 to Swift 3.0. Xcode 8.2 will be the last release that will support Swift 2.3 so we should migrate quickly to Swift 3.0.

jPaolantonio commented:
Is there anything I can do to help the Swift 3 migration?
